### PR TITLE
Use object_size for heap objects

### PR DESCRIPTION
**Pull request by kroening**
This should be performance benchmarked.

### CODE REVIEW

**Comment by tautschnig** [comment]
AFAIK depends on #2852?

**Comment by kroening** [comment]
Yes; #2852 is merged and this is rebased, so this now needs to get performance benchmarking.

File: src/analyses/goto_check.cpp [hunk 1]
```diff
@@ -1152,11 +1141,7 @@ void goto_checkt::bounds_check(
     const exprt &pointer=
       to_dereference_expr(ode.root_object()).pointer();
 
-    if_exprt size(
-      dynamic_object(pointer),
-      typecast_exprt(dynamic_size(ns), object_size(pointer).type()),
-      object_size(pointer));
-
+    auto size=object_size(pointer);
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Nit pick: `const` and clang-format spacing

File: src/analyses/goto_check.cpp [hunk 2]
```diff
@@ -1031,27 +1031,16 @@ goto_checkt::address_check(const exprt &address, const exprt &size)
         not_exprt(dead_object(address, ns)), "dead object"));
     }
 
-    if(flags.is_unknown() || flags.is_dynamic_heap())
-    {
-      const or_exprt dynamic_bounds_violation(
-        dynamic_object_lower_bound(address, ns, nil_exprt()),
-        dynamic_object_upper_bound(address, ns, size));
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp:1039]: To be perfectly clean, shouldn't all the code implementing `dynamic_object_*_bound` be removed as well?

**Review by tautschnig — COMMENTED** [review]
Looks ok to me, but I don't really understand why there is concern about performance, and hence feel I'm missing something. (Thus not yet approving.)

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 306948beb95efd5bbc3e70ab8571407b24ef12b3).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Comment by kroening** [comment]
Performance could be affected as follows: We currently use nondeterminism to determine the object that is out of bounds. This is linear in the number of object accesses and constant in the number of accesses. By contrast, object_size() is linear in the number of objects. I have not yet managed to come up with a way of encoding object_size() using nondeterminism.

File: src/analyses/goto_check.cpp (hunk 1 above)
Comment (kroening) [reply src/analyses/goto_check.cpp]: ok

File: src/analyses/goto_check.cpp (hunk 2 above)
Comment (kroening) [reply src/analyses/goto_check.cpp:1039]: Will do in separate PR.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 836409a6ec64b1cc8b6a4f11992b7af04bcdee31).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84674538

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: de83abb27b70896e009633823c0d034734f7ff1c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84801271

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 188dbeff4760b515dd918edf94773dee864ae999).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86665883

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 3024bda6b5e48a697883c41f51e86e2c2352e6cf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86905932

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 9c938b17e72ee343d89615f50c6fb8b1e407d042).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87885388

**Comment by TGWDB** [comment]
Closing due to age (no further comment on PR content), please reopen with rebase on develop if you intent to continue this work.

**Pull request closed without merging**